[llvm][InstCombine] Fold select to cmp for weak and inverted inequalities

yashnator opened the pull request:
Currently:

```cpp
int signum(int x) {
    if (x < 0) return -1;
    if (x > 0) return +1;
    return 0;
}
```

is not identified as `scmp(x,0)`.

This patch adds folding for the edge case: `(x > -1) ? zext(x != 0) : -1`, which is generated for the above signum and is equivalent to `scmp(x, 0)`. For other constants/variables, the fold is already optimised.

[Alive2 proof](https://alive2.llvm.org/ce/z/EYQZwV) (taken from issue)

Resolves #143259

yashnator committed:
[Instcombine][Tests] Add test for folding select
to cmp with constants in non-canonical inequalities

el-ev reviewed (commented):
Is there any test for the change?

el-ev commented:
> @el-ev I had added a test in scmp.ll- `scmp_x_0_from_gt_minus_1`

Could you please add additional test cases and pre-commit the tests?

Please check https://llvm.org/docs/InstCombineContributorGuide.html

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -3603,6 +3603,18 @@ Instruction *InstCombinerImpl::foldSelectToCmp(SelectInst &SI) {
        ICmpInst::getSwappedPredicate(ExtendedCmpPredicate) == Pred))
     Replace = true;
 
+  // Handle the edge case (x > -1) ? zext(x != 0), -1
```
This can also occur with other constants, not just -1 / 0: https://alive2.llvm.org/ce/z/TIceGg

yashnator replied:
I modified the code to handle weak inequalities and inverted inequalities. So, all constants should work now

I previously handled only 0 because the code generated by C only showed a difference for 0

nikic commented on llvm/test/Transforms/InstCombine/scmp.ll:
```diff
@@ -473,3 +473,17 @@ define i8 @scmp_from_select_eq_and_gt_neg3(i32 %x, i32 %y) {
 }
+
+; Fold (x > -1) ? zext(x != 0), -1 to scmp(x, 0)
+define i32 @scmp_x_0_from_gt_minus_1(i32 noundef %0) local_unnamed_addr #0 {
```
```suggestion
define i32 @scmp_x_0_from_gt_minus_1(i32 %0) {
```

topperc commented:
> Irrelevant to this PR in itself, but I did notice when running both src in tgt in llc, but why does llc lower the src in the alive2 posted better than the intrinsic?

I'm guessing you mean on X86. It looks neutral on RISC-V and scmp looks better for AArch64.

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -3603,6 +3603,18 @@ Instruction *InstCombinerImpl::foldSelectToCmp(SelectInst &SI) {
     Replace = true;
 
+  // Handle the edge case (x > -1) ? zext(x != 0), -1
+  if (IsSigned && ICmpInst::isGT(Pred) && match(FV, m_AllOnes()) &&
```
RHS is not checked.
Miscompilation reproducer: https://alive2.llvm.org/ce/z/2p2mS6
```
; bin/opt -passes=instcombine test.ll -S
define i32 @src(i32 noundef %0) {
  %2 = icmp ne i32 %0, 0
  %3 = zext i1 %2 to i32
  %4 = icmp sgt i32 %0, 1
  %5 = select i1 %4, i32 %3, i32 -1
  ret i32 %5
}
```
```

----------------------------------------
define i32 @src(i32 noundef %#0) {
#1:
  %#2 = icmp ne i32 noundef %#0, 0
  %#3 = zext i1 %#2 to i32
  %#4 = icmp sgt i32 noundef %#0, 1
  %#5 = select i1 %#4, i32 %#3, i32 4294967295
  ret i32 %#5
}
=>
define i32 @tgt(i32 noundef %#0) {
#1:
  %#2 = scmp i32 i32 noundef %#0, 0
  %#range_0_%#2 = !range i32 %#2, i32 4294967295, i32 2
  ret i32 %#range_0_%#2
}
Transformation doesn't verify!

ERROR: Value mismatch

Example:
i32 noundef %#0 = #x00000000 (0)

Source:
i1 %#2 = #x0 (0)
i32 %#3 = #x00000000 (0)
i1 %#4 = #x0 (0)
i32 %#5 = #xffffffff (4294967295, -1)

Target:
i32 %#2 = #x00000000 (0)
i32 %#range_0_%#2 = #x00000000 (0)
Source value: #xffffffff (4294967295, -1)
Target value: #x00000000 (0)
```

yashnator replied:
Thanks for the report! I missed it, my bad. I handled more general inequalities and added this as a test case. It should work now.

AZero13 commented:
> > Irrelevant to this PR in itself, but I did notice when running both src in tgt in llc, but why does llc lower the src in the alive2 posted better than the intrinsic?
> 
> I'm guessing you mean on X86. It looks neutral on RISC-V and scmp looks better for AArch64.

Yeah, might have to work on that.

I mean human written asm for this in particular is:

  add edi, edi
  setnz cl
  sbb eax, eax
  or al, cl
  ret
  
  but I do not know if I can get the compiler to lower that to this.

topperc commented:
> > > Irrelevant to this PR in itself, but I did notice when running both src in tgt in llc, but why does llc lower the src in the alive2 posted better than the intrinsic?
> > 
> > 
> > I'm guessing you mean on X86. It looks neutral on RISC-V and scmp looks better for AArch64.
> 
> Yeah, might have to work on that.
> 
> I mean human written asm for this in particular is:
> 
> ```
>   add edi, edi
>   setnz cl
>   sbb eax, eax
>   or al, cl
>   ret
> ```
> 
> but I do not know if I can get the compiler to lower that to this.

`add edi, edi` seems wrong there

AZero13 commented:
> add edi, edi seems wrong there

Why?

topperc commented:
> > add edi, edi seems wrong there
> 
> Why?

Nevermind, I think its ok, but there's one partial register write and one false dependency in that code.

yashnator commented on llvm/test/Transforms/InstCombine/select-to-cmp.ll:
```diff
@@ -0,0 +1,254 @@
+}
+
+; Fold (x < y + 1) ? sext(x < y) : 1 to ucmp(x, y)
+define i32 @ucmp_x_y_from_lt_add_sext_lt(i32 %x, i32 %y) {
```
I see. Canonical patterns handle variables, but not constants. I fix this

dtcxzyw replied:
Unfortunately the overflow check seems still fragile....

I would suggest you just handle the original motivating case `signum(x) -> scmp(x, 0)`. Then leave the generalization as a follow-up if we can confirm it is profitable in real-world code.

AZero13 commented:
> > > add edi, edi seems wrong there
> > 
> > 
> > Why?
> 
> Nevermind, I think its ok, but there's one partial register write and one false dependency in that code.

partial register write, yes, deliberate
false dependency? Where?

AZero13 commented:
if you're looking at `sbb eax, eax`, then (1) that's probably special-cased by the CPU, and (2) even if not, `eax` isn't touched before that point, so it's very unlikely that the value is still in flight

I don't know if LLVM is aware of special cases like that though. Or if it's worth teaching LLVM that

nikic reviewed (commented):
We have a getFlippedStrictnessPredicateAndConstant() helper for cases like this. I think we can handle this generically by doing a swap of the select arms to get the constant into TV, inverting the predicate, and then using getFlippedStrictnessPredicateAndConstant() to convert the now non-strict predicate into a strict predicate.

topperc commented:
> if you're looking at `sbb eax, eax`, then (1) that's probably special-cased by the CPU, and (2) even if not, `eax` isn't touched before that point, so it's very unlikely that the value is still in flight
> 
> I don't know if LLVM is aware of special cases like that though. Or if it's worth teaching LLVM that

For a long time it was only special cased on AMD CPUs. I don't if Intel finally fixed it or not.

yashnator committed:
[InstCombine] Fold select to cmp with constants in non-canonical inequalities

For constant y, inequalities aren't always canonical
so we need to check the conditions such as
- (x > y - 1) ? zext(x != y) : -1
- (x > y - 1) ? zext(x > y) : -1
- (x < y + 1) ? sext(x != y) : 1
- (x < y + 1) ? sext(x < y) : 1
and similary non-strict equalities.

Fold select into scmp/ucmp based on signedness of
the comparison predicate.

Resolves #143259

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -3572,6 +3572,28 @@ Instruction *InstCombinerImpl::foldSelectToCmp(SelectInst &SI) {
+    std::swap(TV, FV);
+  }
+
+  if (ICmpInst::isGE(Pred) || ICmpInst::isLE(Pred)) {
```
```suggestion
  if (ICmpInst::isNonStrictPredicate(Pred)) {
```

yashnator commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -3632,6 +3632,55 @@ Instruction *InstCombinerImpl::foldSelectToCmp(SelectInst &SI) {
+      match(TV, m_ZExt(m_c_ICmp(ExtendedCmpPredicate, m_Specific(LHS),
+                                m_Value(EXT_RHS)))) &&
+      (ExtendedCmpPredicate == ICmpInst::ICMP_NE ||
+       ICmpInst::isGT(ExtendedCmpPredicate))) {
```
I have modified the code according to Nikita's suggestion, and added this as a test case

nikic reviewed (approved):
LGTM

yashnator commented:
I don't have commit access, can you help merge?
